finance_dl/discover.py

```diff
@@ -0,0 +1,172 @@
+    def __init__(self, credentials, output_directory,
+                 **kwargs):
+        """
+        @param earliest_history_date: Earliest UTC date for which to retrieve
+            transactions and balance information.
+
+        @param max_history_days: Number of days of history to retrieve, starting
+            from the previous UTC day, if `earliest_history_date` is not
+            specified.
+        """
+        super().__init__(**kwargs)
+        self.credentials = credentials
+        self.output_directory = output_directory
+        os.makedirs(self.output_directory, exist_ok=True)
+
+    def check_after_wait(self):
+        check_url(self.driver.current_url)
+
+    def find_account_last4(self):
+        return self.driver.find_element_by_xpath(XPATH_OF_LAST_FOUR_DIGITS).text
+
+
+    def login(self):
+        try:
+            account = self.driver.find_element_by_xpath(XPATH_OF_LAST_FOUR_DIGITS)
+            logger.info("Already logged in")
+        except NoSuchElementException:
+            logger.info('Initiating log in')
+            url = "https://portal.discover.com/customersvcs/universalLogin/ac_main"
+            self.driver.get(url)
+
+            (username, password), = self.wait_and_return(
+                self.find_username_and_password_in_any_frame)
+            logger.info('Entering username and password')
+            username.clear()
+            password.clear()
+            username.send_keys(self.credentials['username'])
+            password.send_keys(self.credentials['password'])
+            with self.wait_for_page_load(timeout=30):
+                password.send_keys(Keys.ENTER)
+            self.check_after_wait()
+            logger.info('Logged in')
+            account = self.wait_and_return(self.find_account_last4, timeout=7)[0]
+        if len(account) != 4:
+            raise Exception("{}: Either identified tag is not the account number, or they've changed its format.".format(account))
+        self.account_name = account
+
+    def download_ofx(self):
+        # Downloads all transactions in current calendar year
+        this_year = datetime.date.today().year
+        start_date = datetime.datetime.strptime(str(this_year), '%Y')
+        end_date = datetime.datetime.strptime(str(this_year + 1), '%Y')
+        logging.info('Fetching history for all of {}'.format(this_year))
+
+        stdt = start_date.strftime('%Y%m%d')
+        enddt = end_date.strftime('%Y%m%d')
+        fname = "Discover-{}.ofx".format(this_year)
+        # Example URL:
+        # https://card.discover.com/cardmembersvcs/ofxdl/ofxWebDownload?stmtKey=W&startDate=20191014&endDate=20191014&fileType=QFX&bid=9625&fileName=Discover-RecentActivity-20191014.qfx
+        qfx_url = "https://card.discover.com/cardmembersvcs/ofxdl/ofxWebDownload?stmtKey=W&startDate={}&endDate={}&fileType=QFX&bid=9625&fileName={}"
+        qfx_url = qfx_url.format(stdt, enddt, fname)
+        logging.info("Downloading from URL: {}".format(qfx_url))
```

Comment:
If you have the url, instead of using get_downloaded_file, another option is to specify `use_seleniumrequests=True` and then use `self.driver.request('GET', url)` --- see paypal.py for an example.

Reply:
Done.